content/blog/2016-07-23-pull-does-not-scale-or-does-it.md

```diff
@@ -0,0 +1,178 @@
+handled similarly. The resulting monitoring traffic is low, and the pull-based
+approach also does not create problems in this case.
+
+## But now my monitoring needs to know about my service instances!
+
+With a pull-based approach, your monitoring system needs to know which service
+instances exist and how to connect to them. Often people are worried about the
+extra configuration this requires on the side of the monitoring system.
+
+We would argue that you cannot escape this configuration effort for
+serious monitoring setups in any case: if your monitoring system doesn't know
+what the world *should* look like and which monitored service instances
+*should* be there, how would it be able to tell when an instance just never
+reports in, is down due to an outage, or really is no longer meant to exist?
+This is only acceptable if you never care about the health of individual
+instances at all, like when you only run ephemeral workers where it is
+sufficient for a large-enough number of them to report in some result. Most
+environments are not exclusively like that.
+
+If the monitoring system needs to know the desired state of the world anyway,
+then a push-based approach actually requires *more* configuration in total. Not
+only does your monitoring system need to know what service instances should
+exist, but your service instances now also need to know how to reach your
+monitoring system. This is not only more required configuration, it also makes
+your monitoring setup less flexible. Now you cannot just run a copy of
+production monitoring on your laptop, as you can with pull. Now you cannot just
+get metrics from some other tool or inspect metrics endpoints manually, as you
+can with pull. Now you cannot just run two Prometheus servers to get high
+availability, as you can with pull.  Now you cannot change the endpoint under
+which your production monitoring is reachable without reconfiguring all of your
+metrics sources.
+
+On a practical front, Prometheus makes it easy to configure the desired state
+of the world with its built-in support for a wide variety of service discovery
+mechanisms for cloud providers and container-scheduling systems: Consul,
+Marathon, Kubernetes, EC2, DNS-based SD, Azure, Zookeeper Serversets, and more.
+Prometheus also allows you to plug in your own custom mechanism if needed.
+In a microservice world or any multi-tiered architecture, it is also
+fundamentally an advantage if your monitoring system uses the same method to
+discover targets to monitor as your service instances use to discover their
+backends.
+
+## Accidentally DDoS-ing your monitoring
+
+Whether you use a pull-based or a push-based approach, any time-series database
+will fall over if you send it more samples than it can handle. However, we have
+made the experience that it's slightly more likely for a push-based approach to
+accidentally bring down your monitoring. If the control over what metrics get
```

Comment:
Although mitigation is the focus, I'd say it's still slightly less likely to happen to begin with because some random intern's first-day test job cannot accidentally send metrics to your monitoring system at all without someone configuring the monitoring system to do so.